### ISSUE TITLE
Graphics overhaul

### ISSUE THREAD

**Issue by dataf3l** [colobot/colobot#1304]
One of the students I teach programming using colobot (for fun, not for profit, it's just too damn fun), told me "do you imagine if this had Fornite Graphics!", I guess the current youth is exposed to high quality games and therefore find the graphics of the game unsatisfactory (I ask her "have you even SEEN logo? this is NICE graphics!", but she wasn't convinced.

So I was wondering, what would it take to improve on the graphics of the game?
Would that require a change in engine? to something like Unity or Unreal?

I'm just asking in behalf of the students, me personally I think the graphics are fine.

**Comment by tomaszkax86** [comment]
Colobot uses Phong reflection model, which uses diffuse, specular and ambient components and adds them together. It's an old model that is no longer used in modern games because it's too simplistic. Modern games use Physically Based Rendering (PBR) approach, usually based on metallic-roughness workflow, which allows far more options to be set and is a much better approximation of reality.

In order to keep graphics consistent, Colobot keeps its OpenGL implementation as close as they can be to the original. It is, however, possible to implement PBR materials in the game. We're considering doing that but it's not high on the priority list right now, mostly because this requires reworking all the models.

**Comment by dataf3l** [comment]
Is adding grass a Possibility ?

**Comment by tomaszkax86** [comment]
Maybe in the future. Implementing grass isn't particularly difficult but it's a problem from level-design perspective.

**Comment by tomangelo2** [comment]
The graphics needs work, both engine (like explained above) and assets are outdated. In result it still looks mostly like original game from 2001 year.
But on the other hand, I don't think we'll be aiming on Fortnite-like design, instead we'll try to modernize and develop vanilla design, both by upgrading current assets and making new ones fitting in with other assets.

**Comment by MrSimbax** [comment]
> I guess the current youth is exposed to high quality games and therefore find the graphics of the game unsatisfactory

This would imply that Colobot is not high quality because of its graphics. How mean! (Just kidding. Or am I?...).

I find this expectation a little troublesome to be honest, especially that this is an open source project and we do not have the resources of a triple A game studio like the guys who are behind Fortnite. I think the youth (in fact, even some adults) should be more exposed to small indie or even retro games to normalize those expectations. Just because a game doesn't use PBR, RTX or other fancy-schmancy resource-consuming graphics techniques doesn't make it a bad game. For example, look at Dwarf Fortress: it's considered one of the most complicated video games ever (you might even argue it's a simulation) and in the vanilla version all you have in terms of graphics are colorful text symbols. Would fancy graphics improve the game? Maybe, but it didn't stop the game from being successful, because the devs focused on the things that mattered for the game the most first and only recently started working on the graphical improvements.

Obviously, you can't have a game without graphics (even text games have graphics, they have to show the text somehow), but it's only one component of the whole and personally I think the fun factor is much more important than graphics. Graphics can significantly contribute to this, it's hard to argue with that, but it's important to remember that the first and foremost goal of video game graphics is to communicate clearly to the player what is happening on the screen. (In that regard, I think Colobot has high-quality graphics actually, some games could learn from it.) What's the point of ultra realistic or beautiful graphics if you quickly discover under the hood that the whole thing is very shallow? Yeah, it can be *very* impressive on the first sight, but it quite quickly becomes boring if the "real" game hidden under all those aesthetics is nothing interesting.

Have you noticed how modern AAA games usually give some kind of "detective sight" ability to players that highlights the interactive elements? Is it a sign that modern games are more shallow because of that? Not really, it's a consequence of the highly detailed environments in games. They are much more detailed than the game world is able to simulate and it's close to impossible to distinguish the interactive elements from the static ones.

With that in mind, let's assume for a moment that we can enhance the Colobot graphics in pretty much any way we want. Keeping the above example, it's clear that it may require adding some new gameplay features and significantly contribute to some design decisions. Currently, the graphics is quite simple and communicates everything clearly. You can recognize an Alien or a Titanium from miles away. The buildings are huge, unrealistic but it's not hard to guess what they do just by looking at them. Nothing is hidden in grass or behind a lot of graphical details. Add too much grass and suddenly the game becomes harder in arguably unfun way: you now have to look harder to find the small TitaniumOres. Colobot is not a game where looking for a needle in a haystack would be a good experience. And that's just grass! Once we add grass, people will probably expect more plants, maybe even jungle-like environments. Now that's just adds tons of complexity to the game and would require a lot of changes in level design. Think about the infamous goto() function from CBOT to not search far. It's unreliable already, now add to it the complexities of modern game environments...

To sum up, I guess the point is, there's always a trade-off in video games (and software in general). If not technical trade-off, then resource trade-off (we have to choose on what to work on because we're only humans with limited lifespan and our own lives). One always has to consider if achieving something non-trivial is really worth the effort for the project. Especially in game design, it's really important to ask yourself: should this feature be there? Surprisingly often, it shouldn't. Even in the hypothetical situation where you actually have unlimited resources! Usually, there are simply more important things to work on for the game before a particular feature should be even considered.

I hope the above explanation will help you convince students that Colobot is fine. It certainly could use some graphical improvements (many, in fact), but considering how old this game is and the situation we're in, I think it really is fine and it makes me sad that someone would think otherwise, regardless of age. Comparing Colobot to huge games like Fortnite is simply not fair. It would make more sense to compare it to other open source games, like OpenTTD.

> So I was wondering, what would it take to improve on the graphics of the game?

A lot of resources. Human resources and time resources. We don't have those. That's unfortunate, but it is how it is. Anyway, before (if) we get to the graphics level of Fortnite the youth will have already forgotten about it and moved on to the Next Big Thing™. It's probably already happening.

> Would that require a change in engine? to something like Unity or Unreal?

Colobot currently uses a custom engine and we can't change that due to legal reasons as far as I'm aware. We (anyone in fact) could make a spiritual successor on a modern game engine (in fact, some people long time ago even started such projects, but that was before Colobot was open sourced) but it would no longer be Colobot. Besides, changing the engine is once again a lot of work so it would probably be infeasible for us anyway. Remember that AAA games such as Fortnite and even some indie games are made by people who spend at least 8 hours a day on such projects. Compare it to Colobot, it's a miracle when someone on the team not only finally has a free day but also decides to use it for working on Colobot specifically. There's not a lot you can do in one day, and the fact that you'll almost certainly have long breaks between each working session doesn't exactly help productivity either.

**Comment by zielmicha** [comment]
> Colobot currently uses a custom engine and we can't change that due to legal reasons as far as I'm aware

Technically Colobot could use Godot, which is on MIT license which is compatible with GPL. Of course, it would take *ton* of work to port (possibly it would be easier to just rewrite the game from scratch...).

**Comment by tomangelo2** [comment]
There is still a thing about final code being derivative of original code, as written:
> EPSITEC SA grants the TerranovaTeam the right to use names Colobot, CeeBot, BuzzingCars and Blupi in association with products based on or derived from the original Colobot source code distribution

So unless taking a part of code and putting it as code in game made in Godot engine or something else counts as product derived it's risky.

**Comment by immibis** [comment]
@tomangelo2 that is literally what a derived product is

**Comment by dataf3l** [comment]
I was researching on this Godot Engine, I hadn't heard of it before (I've used Unity and other engines in the past). 
I think Godot makes a lot of sense, since the graphics are stunning:

https://www.reddit.com/r/godot/comments/ezru4g/open_landscape_demo_in_godot_for_free_use/

**Comment by Erihel** [comment]
Using Godot Engine would mean creating Colobot from scrach.

**Comment by dataf3l** [comment]
One has to wonder, if these ideas:
https://www.youtube.com/watch?v=lctXaT9pxA0&feature=youtu.be
could be applicable to colobot.
Maybe all the project needs is a "Shader person"?

**Comment by tomaszkax86** [comment]
> One has to wonder, if these ideas:
> https://www.youtube.com/watch?v=lctXaT9pxA0&feature=youtu.be
> could be applicable to colobot.
> Maybe all the project needs is a "Shader person"?

I am the "shader person" in the project. Colobot needs its engines rewritten and abandon the GL14 engine before we can do cooler stuff. This will eventually happen.

**Comment by dataf3l** [comment]
I'm very happy to hear we have Shader person.
Which Engines have you considered in addition to Godot?
What are your thoughts on Unity3D ?

**Comment by tomaszkax86** [comment]
By engines I meant OpenGL implementations that the game uses, which are GL14, GL21 and GL33. Moving to completely different game engines, like Godot or Unity3D, would require massive rewriting which is not something we're willing to do right now.

**Comment by HybridDog** [comment]
As far as I know, the radiosity method could be used to precalculate lightmaps with complex global lighting (e.g. indirect lighting), and then during playing, the game would just apply these lightmaps to the albedo textures. This works for static light sources, such as a sun light direction which does not change over time, and static mesh data, so the moving player would need to be lighted separately. I think [Cube2 Sauerbraten](http://sauerbraten.org/) does something like this. In my opinion the rendering in Sauerbraten is very fast although the graphics look complex.

**Comment by immibis** [comment]
IMO with a good architecture it should be possible to keep GL14 but without the cool new graphics. What obstacles do you envision running into?

**Comment by Erihel** [comment]
Having GL14 is pointless. Why waste resources for something that's fixed pipeline and **very** old?

**Comment by tomaszkax86** [comment]
You can't use modern rendering methods with legacy OpenGL. Take PBR workflow, for example, albedo textures can't generally be used as diffuse textures, they are usually completely different. So if we make a switch to PBR, GL14 engine simply won't be able to work. If we want them to continue working, we'll have to deliberately keep legacy textures and separate rendering pipeline, meaning additional work at maintaining them.

**Comment by MrSimbax** [comment]
Do you think we can at least safely bump the default OpenGL version from 1.4 to 2.1 for the next release, which will happen probably before the major graphics engine changes (possibly by introducing a framework like suggested here https://github.com/colobot/colobot/issues/1306#issuecomment-653643108)?

**Comment by tomangelo2** [comment]
IIRC some Radeon GPUs had issues with gl21, but I agree that we should bump the default engine. Maybe with a switch in Device tab to allow to easily downgrade engine in case of unencountered serious issues. Making newer engine as default may help detecting and ironing out such cases.

**Comment by tomaszkax86** [comment]
We should leave version bumping to after graphics overhaul, when newer engines work correctly and better than the current default engine.

**Comment by tomaszkax86** [comment]
The plan is to do complete overhaul and target OpenGL 3.3 as the main rendering engine with OpenGL 2.1 as an option for older hardware. Support will be dropped for the engine that uses OpenGL 1.4.

**Comment by dataf3l** [comment]
I know rewriting from scratch is not on the table, 

however, with the advent of new tools like chatgpt and GitHub pilot, is it possible that learning materials will have to be adapted to the new reality that the new programmers will have to face?

should we use this kind of tools to write the unity 3-D port?

would writing the unity 3-D port be now a fast and easy challenge given that we have these Super Duper tools in our toolbelts?

I understand that it's possible to use these tools to write "3-D graphics Math" code with relatively low cognitive overhead. 

With these new tools at hand, we should be emboldened to take on bigger challenges than before.

I think careful consideration should be given to the fact that we now have new tools, and things are easier to code.

we should consider how the use tools is going to impact the material (they lessons themselves).

how this tools are going to affect the way we teach coding in general.

And how these tools are going to affect how we write the tools that teach programming themselves.

we should not "lie" to children and tell them that this things don't exist 

when, instead, we should teach them how to use them, and also teach them their pitfalls, like hallucination.


maybe an additional lesson explaining to them how to use the tool like gpt in the first place?

this poses a new problem, because most likely GitHub copilot is not going to like the made up language that colobot uses, unfortunately.

so, does that mean that we need to have a language which is compatible with copilot? I.e. Javascript ? Python?

I'm just here is an external observer, asking what I hope are interesting questions, 

I would really like your opinion on these.

**Comment by tomangelo2** [comment]
Unity and Godot are quite uncertain because of legal reasons, also even with ChatGPT and/or Pilot it's tons of work to get to the point we're now. On the other hand, won't these tools help make C++ code?

About in-game language, there were some talks about replacing current CBot interpreter with something more universal, that would make it possible to write not only CBot interpreter, but also other languages. On the other hand, that's a job for a different time, as there are more big projects going on with the engine.

**Comment by immibis** [comment]
Do we have specific requests for graphics? I assume they do not just want more triangles per model, because that is only a minor improvement. Is it lighting? Reflections? Grass? Particle effects?

I never saw Fortnite, but I watched my brother play League of Legends, and I suspect the perception of graphics quality in that game comes down to the particle effects. Every action makes a bunch of unique particle effects giving a feeling of "things are happening" on-screen and that there are many different graphical effects.

**Comment by dataf3l** [comment]
For the current code, I would suggest the following experiment:
```
for each file:

  for each function inside each file:
  
    ask ChatGPT: here is the current code, please explain in pseudocode what it does., if this is not graphics code, SKIP
    ask ChatGPT, given this pseudocode implementation, what features could be added to improve this part of the code, in particular the graphics quality, which do not involve changing other parts of the code, and can be done in isolation by changing only this function.
    ask ChatGPT: change the code, and add a unit test please.

```

then, laboriously go through the process of reading, reviewing and modifying the code as per the new ChatGPT code so that it has the new features it comes up with.

It's a for loop, I can donate the compute by running it on my OpenAI API Key if it "too much api calls, that cost money", if provided a python program that parses the C++ and tries to improve each part, as a donation to the project.

maybe a graphics-code-improver script basically.

also whatever ideas it comes up with we can add as issues so that newcomers can simply find a very specific issue with code location and obejctive and they can contribute.

@immibis are you down to write that code? do you think that would help?

**Comment by dataf3l** [comment]
This is an SCC report on the source code:

```
───────────────────────────────────────────────────────────────────────────────
Language                 Files     Lines   Blanks  Comments     Code Complexity
───────────────────────────────────────────────────────────────────────────────
C Header                   367     66182     8526     20211    37445       1130
C++                        302    153604    24447     12155   117002      28529
CMake                       26      2515      282       467     1766         75
Plain Text                  22      1391      276         0     1115          0
Shell                       12       252       31        32      189         32
Markdown                     9       744      204         0      540          0
gitignore                    6        67       10        12       45          0
Autoconf                     2      2404      545      1577      282          0
C                            2      3426      118       529     2779        413
Python                       2       913       11        28      874         26
YAML                         2       125        6         5      114          0
Objective C                  1        86       20         0       66         11
SVG                          1       402        1         1      400          0
XML                          1        38        6         0       32          0
───────────────────────────────────────────────────────────────────────────────
Total                      755    232149    34483     35017   162649      30216
───────────────────────────────────────────────────────────────────────────────
Estimated Cost to Develop (organic) $5,667,731
Estimated Schedule Effort (organic) 26.589464 months
Estimated People Required (organic) 18.937194
───────────────────────────────────────────────────────────────────────────────
Processed 7661173 bytes, 7.661 megabytes (SI)
───────────────────────────────────────────────────────────────────────────────
```

 find . | grep cpp$| xargs cat | wc
  145683  416028 4490922
  
  
  as you can see, it's not that much code, so it can in fact be fed to ChatGPT, bit by bit.
  if a function is too long, we can skip it too, that will make the program suggestions less complete, but also, we can finish faster.

**Comment by dataf3l** [comment]
Hi, even though "relatively strong words" are being used, I understand your harsh criticism of the idea, to come from a place of love, and where you *really care* about the quality and you really care about the correctness of software.

I think correct software is important.

I am not suggesting that you let the computer "loose" on the code with zero supervision.

I am merely suggesting to automate the generation of specific suggestions based on code snippets.

wether or not it should be *the computer* implementing those suggestions or it should be *a human implementing it with the help of the computer*, that would be the decision of the person doing the job.

In my opinion ChatGPT brings a great opportunity to automate the improvement of code, massively, and automatically.

it may even suggest new features for the parser or simplified logic for the parser of code, or better error messages for the students for the educational parser.

I think that the success of the idea really depends on the quality of the *prompt engineering* done before large scale automation, clearly, as you suggest by your comment, a small and flimsy prompt simply won't do.

As for the problem of how to keep consistency between changesets I think basically the problem might be basically to give it an example of the old code and the new code so that the computer can kind of know exactly what we're going for, to provide general rules of thumb, and perhaps even to ask it for multiple suggestions so a human can choose, converting the one-hot problem into a multi-shot problem (multi-shot is easier to achieve).

We can provide example of the old API methods, and also off the new API methods in the prompt, so that it is more likely that the computer will suggest code which uses a new libraries we want to implement, we can increase the likelihood, but this is an statistical endeavor, so maybe it won't be correct 100% of the time.

We can also provide code guidelines to direct the computer, as to the usage of new language features (like new C++ stuff), as I understand it the language has changed a lot in the past 20 years these days much more ergonomic to use, and the new code is usually shorter and more readable.

We can probably also instruct the computer to write generic code with perhaps better type guarantees in order to increase the level of abstraction of the solution and the type correctness.

I think the problem of having two separate prompts in two separate chatGPT sessions, that change to different parts of the code, without creating a conflict between changesets, is in fact an interesting problem, and one worth exploring.

Perhaps a really easy place where the computer can add value is used to ask you to write more internationalization for example i18n into Spanish or other languages like that.

Latin American children also deserve a good education.

I think maybe some kind of long term memory (faiss?) is require providing examples of the changes that were made, 
Imagine a future where we have a database of change-sets,embeddings act as key store using vectors, for example FAISS comes to mind.  or even hnswlib, using this database we can probably ask the computer to change the new code considering examples of previous good change sets that have done before as part of the prompt, we can find all the change sets efficiently using the vector database which will act as a long-term memory for the project.

If the computer is invalid code we can put that inside a for loop check if the new code compiles, if it doesn't we can give the error message back to chatgpt automatically, similar to the gpt-engineer project.

Perhaps the status of suggestions idea can also be used during the actual teaching of the programming to the children, where if their children writes really bad code in a pattern which a lot of people have them before we can have a database of that and we can tell people while this is a common problem and your problem is soon understand XYZ and the solution is to do ABC. So a database of suggestions for children for improvement of the code may actually be a useful idea even if the database is written by humans, we can maybe even use regexen for this.
.

I think it is possible to calculate call graph of the code, and to execute all these tasks in the order of the call graph.

if a new variable needs to be passed around, so the new code parameter can propagate.
we can ask the computer to create tasks for itself for other parts of the code, for example, if I change X i need to remember to change Y in other places, this can be added to the memory after the execution, so it can remember to do that later, on it's own discresion.

(or perhaps, the reverse order of the callgraph), for changes which are more atomic like unit tests addition.

We can probably ask it to ignore sections of the code which are unrelated to the task at hands "graphics improvement".

This, will prevent performing the task on sections of the code which we want o leave alone, like the parser or the language files.

Using the callgraph it will also help the computer do it in a more organized way and also in a way which is logical and step by step.

We can provide it as context *the immediate previous changes* that were performed that were successful and we can also provide it as context the previous changes to the code that were *unsuccessful* as negative examples, this way we can guide the artificial intelligence to provide better code faster.

After the changes are done, we can ask the "code critiquer module" to critique it's own solution so it can improve on it's own work.

Just like you, I also do not believe that just asking the computer to improve the quality of the graphics willy-nilly and letting it loose on the code base will have a good result, with the current version that is, given the limited context, however I do believe that with careful planning it should actually be possible though for the computer to fix it's own code and improve it, which will look like magic when it works.

# Contract between components:

Of course there is the problem of whether or not the contract within components should change, for example if the computer want to improve the code by adding a new parameter to a method, this is most likely going to break the software because it is very likely that the new parameters will need to be spread out all over the place.

However I think there is a lot that can be done even without changing the contract between functions and contract between the elements of the code.

we can also ask it to tell itself to "not forget to add the new param call everywhere where it is required" when new code shows up, which would feed it's own prompt ideas back into itself.

# Pilot

Perhaps a small pilot could be a good idea where the decision is made to change only a very small subset of the code where we think the computer will do a good job at improving (i.e. the graphics), and try out the small pilot to see if it's actually possible to write a specific prompt that will improve the code without breaking it too much the core code.

like maybe writing more stories, adding more ideas to the current stories, creating new and exciting lessons, etc.

# anecdata

Yesterday I had a good experience asking the code to write some LERP code for me (nothing fancy, but it works), if I am very honest I don't think I understand all the code that the computer wrote, and I know I can Google it, but now that the task is finished I am de-incentivized to do it.

So this poses and even more complicated challenge which is that if the computer writes the code and we don't understand it then there is the risk that the solution will become incomprehensible for the humans working on it which is in fact a real risk.

so when the computer writes the code and we don't understand it then when the code breaks then we are going to have a hard time trying to fix it, so with this in mind I would strongly suggest that we *ask ChatGPT in the prompt to write   really simple code* so that the code is really easy to understand even (if it's a little longer and less abstract), so that we can actually in fact continue with the maintenance of the software in the future.

I sincerely hope that you guys reconsider of your position and to try out the idea, I also strongly suggest that the code used to improve the software is also released as open source, mostly because I am very curious to see what it will look like, for my own education.

Considering the fact that the AI is now able to write code, perhaps it would be a good idea to teach children about this fact.

So that they can in fact learn how to prompt the computer to write the code that they actually need in a very precise way, we shall also need educate the children how old the dangers of having a computer write the code and then them not understanding the code.

Perhaps the AI tool could be integrated directly into the software.

For example, if a person writes the code incorrectly then perhaps we can give ChatGPT: the Colobot Langauge spec, the error message and the user code, and also an explanation of the exercise's objective to the artificial intelligence, so that it can provide suggestions to the user on how to fix their code *without actually telling them exactly how to fix the code* or providing code on it's own.

In this scenario the AI would be acting as a patient language tutor at it would be teaching the programmers of the future how to write code which is efficient, fast, it does the job nicely, it's elegant, and most importantly,  children will have the opportunity to interact with their machine to work together with a machine to write a code, because this is what most likely their real life job will look like.

so that we can prepare them for the real life job and not for fantasy world which is pre-pandemic and may be no longer existent where people used to write code on their own, as was the case just a mere couple years ago.

We need to educate them for the future, not the past.

I think we can also ask chad GPT to suggest to children how to improve their variable names, how to make more abstract code, how to make more intelligent, elegant, functional, how to avoid code of duplication, and how to write better comments, because these people are in the very first stages of learning how to write software, I think it actually makes sense to guide them in the right way so that the coding right is actually good quality.

this way the new code by the children won't look too odd to existing devs, and they will be able to work with other people nicely.

I think we have a responsibility towards the new generation, because otherwise in the future we will see all the new developers to write very bad code, without any supervision and we will end up with large months of very bad quality code that we (as the previous generation) will need to fix.

Please consider this.

**Comment by tomaszkax86** [comment]
If you believe you can use ChatGPT to refactor all of Colobot and even make it better, I welcome you to a trivial challenge. Fork Colobot and use ChatGPT to do just that. Feed it old and buggy code so it spits out new and improved code. One by one replace bad parts with new parts. I highly doubt this will work but feel free to prove me wrong.

**Comment by dataf3l** [comment]
Hi, thank you for your prompt response, I don't feel the current code to be particularly old or buggy, I'm just mentioning that the student complained about the graphics, as I said earlier, to me, graphics are ok. this is a User request, not a dev request.

As per your suggestion to fork the code, I would much rather contribute the colobot, than make my own, I believe in community and collaboration, the free exchange of ideas, instead of "splitting off".

All I'm doing is contributing ideas.

**Comment by dataf3l** [comment]
Is there even value in a bug-less version, if nobody participates continuously in the improvement?

**Comment by Erihel** [comment]
ChatGPT is more like a trained language model than an actual intelligence. Authors admit that it can give incorrect answers https://openai.com/blog/chatgpt. It can be used as a helping tool but it won't write complex software by its own.

I understand there's an hype for anything AI driven. Just like they use to add *smart* to anything, even if it was just a blinking led light.

**Comment by dataf3l** [comment]
This is precisely the reason why I am suggesting we ask for multiple possible versions, so that we can choose the best version, or even none at all.

**Comment by immibis** [comment]
If you think it is easy you can do it yourself. It is not easy. It is probably not even possible.

**Comment by dataf3l** [comment]
I work with an M1 Machine so I am probably blocked by this:

https://github.com/colobot/colobot/issues/1564

once this issue is resolved I can try and offer suggestions for improvement.

I never said it was easy, but I don't think the main cognitive cost will be spent by humans, I think it's better to let the machine do the thinking.

Perhaps it can also be used to give performance improvements...

**Comment by dataf3l** [comment]
I think large tasks some times look daunting, but we can Ask ChatGPT to break down the task into smaller subtasks.
alternatively, we can also use our brains, like in the olden days of yore.

**Comment by tomangelo2** [comment]
@dataf3l You've mentioned you're probably blocked by mentioned issue. Do you have same error? If so, can you try to compile with `-DCMAKE_CXX_FLAGS=-Wno-missing-declarations` CMake flag?

**Comment by dataf3l** [comment]
Well, the flag DOES help in fact.

 I tried cmake but it doesn't build, there is a missing "website" variable somewhere, unfortunately:


```

3 warnings generated.
[ 11%] Building CXX object src/CMakeFiles/colobotbase.dir/common/restext.cpp.o
[ 11%] Building CXX object src/CMakeFiles/colobotbase.dir/common/system/system.cpp.o
[ 12%] Building CXX object src/CMakeFiles/colobotbase.dir/common/system/system_other.cpp.o
[ 12%] Building CXX object src/CMakeFiles/colobotbase.dir/common/system/system_macosx.cpp.o
/Users/b/Downloads/colobot-main/src/common/system/system_macosx.cpp:74:5: warning: 'FSFindFolder' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
    FSFindFolder( kUserDomain, folderType, kCreateFolder, &ref );
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Folders.h:286:1: note: 'FSFindFolder' has been explicitly marked deprecated here
FSFindFolder(
^
/Users/b/Downloads/colobot-main/src/common/system/system_macosx.cpp:75:5: warning: 'FSRefMakePath' is deprecated: first deprecated in macOS 10.8 [-Wdeprecated-declarations]
    FSRefMakePath( &ref, reinterpret_cast<UInt8*>(&path), PATH_MAX );
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Files.h:4086:18: note: 'FSRefMakePath' has been explicitly marked deprecated here
extern OSStatus  FSRefMakePath(const FSRef *ref, UInt8 *path, UInt32 pathBufferSize)     __OSX_AVAILABLE_BUT_DEPRECATED(__MAC_10_0, __MAC_10_8, __IPHONE_NA, __IPHONE_NA);
                 ^
/Users/b/Downloads/colobot-main/src/common/system/system_macosx.cpp:138:76: error: use of undeclared identifier 'website'
        GetLogger()->Error("Failed to open website: %s, error code: %i\n", website.c_str(), result);
                                                                           ^
2 warnings and 1 error generated.
make[2]: *** [src/CMakeFiles/colobotbase.dir/common/system/system_macosx.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/colobotbase.dir/all] Error 2
make: *** [all] Error 2
```

a lot of deprecated warnings too.
so I guess that's a different issue.

**Comment by dataf3l** [comment]
Do you have any particular branch you want me to test?

**Comment by tomangelo2** [comment]
Do you have up-to-date dev branch? This variable was fixed already.

**Comment by immibis** [comment]
Can you guys keep the graphics overhaul thread for talking about graphics overhaul and not how to run ChatGPT?

**Comment by dataf3l** [comment]
Sure, no problem, I was able to get it to build at 60%, but there is some weird linker issue.

In other news, perhaps you guys may consider this to be interesting.

https://arxiv.org/abs//2306.17156?utm_source=substack&utm_medium=email

**Comment by immibis** [comment]
For discussion not related to graphics overhaul you can both use the official Discord "server": https://discord.gg/56Fm9kb
